cmd/util/ledger/reporters/fungible_token_tracker.go

```diff
@@ -0,0 +1,201 @@
+	// token type
+	TypeID string `json:"type_id"`
+}
+
+type job struct {
+	owner    flow.Address
+	payloads []ledger.Payload
+}
+
+// Report creates a fungible_token_report_*.json file that contains data on all fungible token Vaults in the state commitment.
+// I recommend using gojq to browse through the data, because of the large uint64 numbers which jq won't be able to handle.
+func (r *FungibleTokenTracker) Report(payloads []ledger.Payload) error {
+	r.rw = r.rwf.ReportWriter(FungibleTokenTrackerReportPrefix)
+	defer r.rw.Close()
+
+	wg := &sync.WaitGroup{}
+
+	// we need to shard by owner, otherwise ledger won't be thread-safe
+	addressCount := 0
+	payloadsByOwner := make(map[flow.Address][]ledger.Payload)
+
+	for _, pay := range payloads {
+		owner := flow.BytesToAddress(pay.Key.KeyParts[0].Value)
+		if len(owner) > 0 { // ignoring FVM values
+			m, ok := payloadsByOwner[owner]
+			if !ok {
+				payloadsByOwner[owner] = make([]ledger.Payload, 0)
+				addressCount++
+			}
+			payloadsByOwner[owner] = append(m, pay)
+		}
+	}
+
+	jobs := make(chan job, addressCount)
+	r.progress = progressbar.Default(int64(addressCount), "Processing:")
+
+	for k, v := range payloadsByOwner {
+		jobs <- job{k, v}
+	}
+
+	close(jobs)
+
+	workerCount := runtime.NumCPU()
+	for i := 0; i < workerCount; i++ {
+		wg.Add(1)
+		go r.worker(jobs, wg)
+	}
+
+	wg.Wait()
+
+	err := r.progress.Finish()
+	if err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func (r *FungibleTokenTracker) worker(
+	jobs <-chan job,
+	wg *sync.WaitGroup) {
+	for j := range jobs {
+
+		view := migrations.NewView(j.payloads)
+		st := state.NewState(view)
+		sth := state.NewStateHolder(st)
+		accounts := state.NewAccounts(sth)
+		storage := cadenceRuntime.NewStorage(
+			&migrations.AccountsAtreeLedger{Accounts: accounts},
+		)
+
+		owner, err := common.BytesToAddress(j.owner[:])
+		if err != nil {
+			panic(err)
+		}
+
+		for _, domain := range domains {
+			storageMap := storage.GetStorageMap(owner, domain)
+			itr := storageMap.Iterator()
+			key, value := itr.Next()
+			for value != nil {
+				r.iterateChildren(append([]string{domain}, key), j.owner, value)
+				key, value = itr.Next()
+			}
+		}
+
+		err = r.progress.Add(1)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	wg.Done()
+}
+
+func (r *FungibleTokenTracker) iterateChildren(tr trace, addr flow.Address, value interpreter.Value) {
+
+	compValue, ok := value.(*interpreter.CompositeValue)
+	if !ok {
+		return
+	}
+
+	if compValue.IsResourceKinded(nil) {
+		typeIDStr := string(compValue.TypeID())
+		hasTargetType := false
+		for _, vt := range r.vaultTypeIDs {
+			if typeIDStr == vt {
+				hasTargetType = true
+				break
+			}
+		}
+
+		if hasTargetType {
+			b := uint64(compValue.GetField(nil, nil, "balance").(interpreter.UFix64Value))
+			if b > 0 {
+				r.rw.Write(TokenDataPoint{
+					Path:    tr.String(),
+					Address: addr.Hex(),
+					Balance: b,
+					TypeID:  string(compValue.TypeID()),
+				})
```

Comment:
it might still include nested vaults? so we do continue iterating, though I removed the need to iterate over the values that are not resource typed, since it won't be possible to have a resource type in a non-resource container.